extensions/interactions/AlgebraicExpressionInput/directives/algebraic-expression-input-rules.service.ts

```diff
@@ -0,0 +1,47 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Rules service for the AlgebraicExpressionInput interaction.
+ */
+
+import { Injectable } from '@angular/core';
+import { downgradeInjectable } from '@angular/upgrade/static';
+
+var nerdamer = require('nerdamer');
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AlgebraicExpressionInputRulesService {
+  MatchesExactlyWith(answer: string, inputs: {x: string}): boolean {
```

Comment:
Why is inputs an object, instead of a string? Is it something that is included in the existing framewok?

Reply:
Yes, that is how the other rules services have implemented this as well. This is because of the way that the input has been interpolated in the rules. For instance, the current rule is `matches exactly with {{x|AlgebraicExpression}}`. So the inputs field is interpolated as an object with the key `x` of type AlgebraicExpression which is a unicode string.

Reply:
True, but can't you change the calling function to pass `inputs.x` in the first place, so the parameters are both strings? Since, I see you have created objects in the parameter in some locations which are not ideal.

Reply:
That wouldn't work since those aren't the only places that the rules functions are being used at. The rules service is being passed to the `CurrentInteractionService` via the `onSubmit` function. This is being done in `oppia-interactive-algebraic-expression-input.component.ts`. Now, this service is general for all interactions and hence requires the inputs parameter to be an object. This is because not all interactions have simple input types like this one, so having them wrapped in an object makes things easier to handle.

Reply:
But `CurrentInteractionService` never calls the function directly? Or does it take the function name (for rule) from interaction_specs and calls it generally?

Reply:
The `CurrentInteractionService` does call the `onSubmit` function. It does this in order to validate the submitted answer against the rules provided in the rules service. And this is done generally which is why all rules need to have the same type of parameter for input. For reference, the function is being called [here](https://github.com/oppia/oppia/blob/89001c40abeb5468c2dd69bdfe8ca24f26982cd0/core/templates/pages/exploration-player-page/services/current-interaction.service.ts#L116).